LocalDOMWindow::getMatchedCSSRules asserts when passed null/empty pseudoElement

jhammer opened the pull request:
#### 9b380bdb5e474414c07bdee71f33357e8ded048f
<pre>
LocalDOMWindow::getMatchedCSSRules asserts when passed null/empty pseudoElement

<a href="https://bugs.webkit.org/show_bug.cgi?id=276677">https://bugs.webkit.org/show_bug.cgi?id=276677</a>

Reviewed by Tim Nguyen.

As of 4d69396, it is no longer valid to pass PseudoId::None to pseudoStyleRulesForElement,
because it triggers an assertion failure in the ctor for PseudoElementRequest. This changes
getMatchedCSSRules to pass std::nullopt instead of PseudoId::None.

Test: Tools/TestWebKitAPI/Tests/mac/GetMatchedCSSRules.mm

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getMatchedCSSRules const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/GetMatchedCSSRules.mm: Added.
(-[GetMatchedCSSRulesTest webView:didFinishLoadForFrame:]):
(TestWebKitAPI::TEST(WebKitLegacy, GetMatchedCSSRulesTest)):

Canonical link: <a href="https://commits.webkit.org/281088@main">https://commits.webkit.org/281088@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30c0304e5a13698fc4df77adaabe290ff9110208

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58671 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37998 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11157 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62298 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9115 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60800 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45635 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9314 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47475 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6489 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60702 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35566 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50734 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28329 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32309 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8032 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8119 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54262 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8307 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64001 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2584 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8305 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54797 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2593 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50761 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54883 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12973 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2172 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33827 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34913 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35997 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34658 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1633,7 +1633,9 @@ RefPtr<CSSRuleList> LocalDOMWindow::getMatchedCSSRules(Element* element, const S
-    auto pseudoId = pseudoElementIdentifier ? pseudoElementIdentifier->pseudoId : PseudoId::None;
+    std::optional<PseudoId> pseudoId;
+    if (pseudoElementIdentifier)
+        pseudoId = pseudoElementIdentifier->pseudoId;
```
Since `pseudoStyleRulesForElement` takes in a `std::optional<Style::PseudoElementIdentifier>`, I'd just pass `pseudoElementIdentifier` (which is already the right type) where we're currently passing the `pseudoId`

nt1m replied:
We can then entirely get rid of the pseudoId variable.

jhammer replied:
Whoops. You are right. Thank you. I will update the pull request.

jhammer replied:
Updated.

nt1m reviewed (approved):
Looks good, but please squash commits into a single one (the first commit message is perfect!), since merge-queue can't land this otherwise.

nt1m commented on Tools/TestWebKitAPI/Tests/mac/GetMatchedCSSRules.mm:
```diff
@@ -0,0 +1,73 @@
+
+namespace TestWebKitAPI {
+
+TEST(WebKitLegacy, GetMatchedCSSRulesTest)
```
Sorry I just noticed this used WebKitLegacy in the boilerplate.

You'll probably should write a WK2 test, and invoke the `window.getMatchedCSSRules` JS API from a user content world like so:

https://searchfox.org/wubkat/rev/7dacb873108c102a5c83ecc6a3c85f361446e678/Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebViewEvaluateJavaScript.mm#199-201

jhammer replied:
No problem. I just updated the PR with a WK2 test.

Commit:
LocalDOMWindow::getMatchedCSSRules asserts when passed null/empty pseudoElement

https://bugs.webkit.org/show_bug.cgi?id=276677

Reviewed by Tim Nguyen.

As of 4d69396, it is no longer valid to pass PseudoId::None to pseudoStyleRulesForElement,
because it triggers an assertion failure in the ctor for PseudoElementRequest. This changes
getMatchedCSSRules to pass std::nullopt instead of PseudoId::None.

Test: Tools/TestWebKitAPI/Tests/mac/GetMatchedCSSRules.mm

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::getMatchedCSSRules const):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/mac/GetMatchedCSSRules.mm: Added.
(-[GetMatchedCSSRulesTest webView:didFinishLoadForFrame:]):
(TestWebKitAPI::TEST(WebKitLegacy, GetMatchedCSSRulesTest)):

Canonical link: https://commits.webkit.org/281088@main
